src/queue.rs

```diff
@@ -0,0 +1,98 @@
+use core::{
+    marker::PhantomData,
+    mem::{size_of, MaybeUninit},
+};
+
+use esp_idf_sys::TickType_t;
+
+use crate::sys;
+
+pub struct Queue<'a, T> {
+    ptr: sys::QueueHandle_t,
+    is_owned: bool,
+    _marker: PhantomData<&'a mut T>,
+}
+
+unsafe impl<'a, T> Send for Queue<'a, T> where T: Send + Sync {}
+unsafe impl<'a, T> Sync for Queue<'a, T> where T: Send + Sync {}
+
+impl<'a, T> Queue<'a, T>
+where
+    // ensures the contained elements are not `Drop`
+    // might be able to lift restriction in the future
+    T: Copy,
+{
+    pub fn new(size: usize) -> Self {
+        Queue {
+            ptr: unsafe { sys::xQueueGenericCreate(size as u32, size_of::<T>() as u32, 0) },
+            is_owned: true,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Create a new queue which is not deleted on `Drop`, but owned by somebody else.
+    ///
+    /// # Safety
+    ///
+    /// Care must be taken that the queue is valid for the constructed
+    /// lifetime.
+    pub unsafe fn new_borrowed(ptr: sys::QueueHandle_t) -> Self {
+        assert!(!ptr.is_null());
+
+        Queue {
+            ptr,
+            is_owned: false,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn as_raw(&self) -> sys::QueueHandle_t {
```

Comment:
I think the general pattern for this method throughout the HAL was to name it `handle`? Can you check with the other drivers that have such a method and change accordingly?

Reply:
The name for the field is `handle`, didn't see any methods called `handle`.